crypto/sm2/sm2_pmeth.c

```diff
@@ -168,21 +158,11 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
 
     switch (type) {
     case EVP_PKEY_CTRL_EC_PARAMGEN_CURVE_NID:
-        group = EC_GROUP_new_by_curve_name(p1);
-        if (group == NULL) {
+        if (p1 != NID_sm2
+            || (group = EC_GROUP_new_by_curve_name(p1)) == NULL) {
             SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_INVALID_CURVE);
             return 0;
```

Comment:
To this end, it probably make sense to add tests that showcase
- [x] how to create an EVP:EC key over the curve defined in the SM2 standard and ensure it can perform EC keygen and compute ECDH and ECDSA sign/verify [I saw you already updated the tests to include this]
- [ ] and a separate tests that showcases how to create an EVP:SM2 key over another arbitrary curve and verify that it does SM2 keygen and SM2 kexchange, sign/verify.

Reply:
> I am not sure about this: doesn't this effectively prevent to execute SM2 keygen and signatures on a different curve than the standard one?

The only thing this prevents is setting an arbitrary curve on an SM2 EVP_PKEY, but since you can switch back and forth between `EVP_PKEY_EC` and `EVP_PKEY_SM2` with `EVP_PKEY_set_alias_type()`, it's perfectly possible to either create a normal EC key with whatever curve you want or switch an existing one to `EVP_PKEY_EC` and do what you need, then switch it to be `EVP_PKEY_SM2` and get SM2 computations.

This is, of course, complete hackery, non-standard stuff, and it *should* have to be done explicitly.

As for SM2 computation over another arbitrary curve, I'm perfectly willing to do that if someone (you?) provides me with a suitable test vector, i.e. a key, input, sm2 id, and signature output.  It should be someone else than me producing that, to avoid being self serving.
There's no need to create a new key, we already have plenty of keys to choose from in evppkey.txt.

Reply:
Dead silence...

Ok, I worked `tesst/ecdsatest.c` a bit to run all curves with "normal" EC computations as well as with SM2 computations.  That should cover everything you asked for.

Reply:
This definitely requires a CHANGES entry! Strictly speaking this is a breaking change - should we allow it? Does this require an OMC decision since it is against policy.

Reply:
...oh but wait....on looking at this a little more deeply it seems that although you could set these parameters to arbitrary curves there was no way to use them because the paramgen and keygen functions were previously NULL. AFAICT the previous `gen_group` field was write only and never used! Am I right or did I miss something. If I'm right maybe this isn't so bad.

Reply:
I can't really sit in front of a computer to run tests and figure all the implications for now sorry!

Hopefully some other reviewer can have a look at it? Apart from the things I mentioned and Richard evaluated I did not have any blocking comment looking at the code on github!

Reply:
Previously, you always generated an EC key and then changed the type to `EVP_PKEY_SM2` with `EVP_PKEY_set_alias_type`, so it was effectively impossible to set the `gen_group` in the SM2 `EVP_PKEY_METHOD`, or use it.

With this PR, generating SM2 keys can still be done as before, *but* now you can also generate SM2 keys following the same pattern as other key types, i.e.:

``` C
    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_SM2, NULL);
    EVP_PKEY *pkey = NULL;

    EVP_PKEY_keygen_init(ctx);
    EVP_PKEY_keygen(ctx, &pkey);
```

Doing it this way *forces* the SM2 curve, which I would say it should.  If you want to generate an EC key with the SM2 curve, you just do as before:

``` C
    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
    EVP_PKEY *pkey = NULL;

    EVP_PKEY_keygen_init(ctx);
    EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx, NID_sm2);
    EVP_PKEY_keygen(ctx, &pkey);
```

So in essence, I haven't changed anything already present, I have just added the possibility to generate SM2 keys directly.

However, looking at the code we're commenting in, I'm wondering why I create a group that I do nothing with.  I'm amazed that hasn't caused a leak yet...

Reply:
Obviously, I disagree regarding this being a breaking change.  I tried to avoid that quite carefully.

Reply:
I changed to not create the group, and added a pretty thorough comment.